File: schemas.py
"""
Validation Schemas for PhenoRx Adverse Drug Reaction Engine
Uses Pydantic V2 for strict type checking, boundary enforcement, and sanitized field coercions.
"""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, field_validator, model_validator


class PatientDataSchema(BaseModel):
    """Demographics, physiological vitals, clinical lab markers, and pharmacogenomic profile"""

    # Demographics
    name: str = Field(default="Patient", max_length=100)
    patient_id: str = Field(default="PT-UNKNOWN", max_length=50)
    age: int = Field(..., ge=0, le=125, description="Patient age in years (0 - 125)")
    sex: str = Field(default="M", description="'M' or 'F'")
    ethnicity: str = Field(default="Asian", max_length=50)

    # Anthropometrics
    height: float = Field(default=170.0, ge=30.0, le=250.0, description="Height in cm")
    weight: float = Field(default=70.0, ge=2.0, le=300.0, description="Weight in kg")
    bmi: float = Field(default=24.2, ge=0.0, le=80.0, description="Body Mass Index (kg/m²)")

    # Renal Panel
    creatinine: float = Field(default=1.0, ge=0.1, le=25.0, description="Serum Creatinine (mg/dL)")
    egfr: float = Field(default=90.0, ge=1.0, le=200.0, description="Estimated GFR (mL/min/1.73m²)")

    # Hepatic Panel
    ast_alt: float = Field(
        default=25.0, ge=1.0, le=5000.0, description="AST/ALT Liver Enzyme Level (U/L)"
    )
    bilirubin: float = Field(default=0.8, ge=0.1, le=50.0, description="Total Bilirubin (mg/dL)")
    albumin: float = Field(default=4.2, ge=0.5, le=7.0, description="Serum Albumin (g/dL)")

    # Complete Blood Count (CBC)
    hemoglobin: float = Field(default=14.0, ge=2.0, le=25.0)
    hematocrit: float = Field(default=42.0, ge=5.0, le=75.0)
    wbc_count: float = Field(default=7000.0, ge=100.0, le=100000.0)
    platelet_count: float = Field(default=250000.0, ge=5000.0, le=2000000.0)
    rbc_count: float = Field(default=4.8, ge=0.5, le=10.0)

    # Differential (%)
    neutrophils: float = Field(default=60.0, ge=0.0, le=100.0)
    lymphocytes: float = Field(default=30.0, ge=0.0, le=100.0)
    monocytes: float = Field(default=7.0, ge=0.0, le=100.0)
    eosinophils: float = Field(default=2.0, ge=0.0, le=100.0)
    basophils: float = Field(default=1.0, ge=0.0, le=100.0)

    # RBC Indices
    mcv: float = Field(default=90.0, ge=40.0, le=150.0)
    mch: float = Field(default=30.0, ge=10.0, le=50.0)
    mchc: float = Field(default=34.0, ge=15.0, le=45.0)
    rdw: float = Field(default=13.0, ge=5.0, le=40.0)

    # Hemodynamics & Vitals
    bp_systolic: float = Field(default=120.0, ge=40.0, le=260.0)
    bp_diastolic: float = Field(default=80.0, ge=30.0, le=180.0)
    heart_rate: float = Field(default=75.0, ge=25.0, le=250.0)
    temperature: float = Field(default=37.0, ge=30.0, le=45.0)

    # Pharmacogenomics - Cytochrome P450 Enzymes
    cyp2c9: str = Field(default="Wild")
    cyp2d6: str = Field(default="EM")
    cyp3a4: str = Field(default="Normal")
    cyp1a2: str = Field(default="Normal")
    cyp2b6: str = Field(default="Normal")
    cyp2c19: str = Field(default="EM")

    # Pharmacogenomics - Transporters
    slco1b1_genotype: str = Field(default="*1/*1")
    abcb1_genotype: str = Field(default="CC")
    abcg2_genotype: str = Field(default="Wild/Wild")

    # Pharmacogenomics - HLA Hypersensitivity Alleles
    hla_a_typing: List[str] = Field(default_factory=list)
    hla_b_typing: List[str] = Field(default_factory=list)
    hla_drb1_typing: List[str] = Field(default_factory=list)

    # Drug Regimen
    medication_name: str = Field(default="Warfarin", max_length=100)
    index_drug_dose: float = Field(default=5.0, ge=0.01, le=10000.0)
    time_since_start_days: int = Field(default=7, ge=0, le=3650)
    concomitant_drugs_count: int = Field(default=0, ge=0, le=50)

    # Clinical Comorbidities (0 or 1)
    diabetes: int = Field(default=0, ge=0, le=1)
    liver_disease: int = Field(default=0, ge=0, le=1)
    ckd: int = Field(default=0, ge=0, le=1)
    cardiac_disease: int = Field(default=0, ge=0, le=1)
    hypertension: int = Field(default=0, ge=0, le=1)
    respiratory_disease: int = Field(default=0, ge=0, le=1)
    neurological_disease: int = Field(default=0, ge=0, le=1)
    autoimmune_disease: int = Field(default=0, ge=0, le=1)

    # Risk Flags (0 or 1)
    cyp_inhibitors_flag: int = Field(default=0, ge=0, le=1)
    qt_prolonging_flag: int = Field(default=0, ge=0, le=1)
    hla_risk_allele_flag: int = Field(default=0, ge=0, le=1)
    inpatient_flag: int = Field(default=0, ge=0, le=1)
    prior_adr_history: int = Field(default=0, ge=0, le=1)
    polypharmacy_flag: int = Field(default=0, ge=0, le=1)
    high_risk_drug_flag: int = Field(default=0, ge=0, le=1)

    # Calculated Dosage Metrics
    cumulative_dose_mg: float = Field(default=0.0, ge=0.0)
    dose_density_mg_day: float = Field(default=0.0, ge=0.0)

    # Extra Medication Context
    external_drugs: List[Dict[str, Any]] = Field(default_factory=list)
    external_drugs_list: List[str] = Field(default_factory=list)
    all_medications: List[str] = Field(default_factory=list)

    @field_validator("sex", mode="before")
    @classmethod
    def sanitize_sex(cls, v: Any) -> str:
        s = str(v).strip().upper()
        if s in ["M", "MALE"]:
            return "M"
        if s in ["F", "FEMALE"]:
            return "F"
        return "M"

    @model_validator(mode="after")
    def compute_derived_metrics(self) -> "PatientDataSchema":
        """Compute automatic BMI if not supplied or zero"""
        if self.bmi <= 5.0 and self.height > 0:
            h_m = self.height / 100.0
            self.bmi = round(self.weight / (h_m * h_m), 1)

        # Polypharmacy flag sync
        if self.concomitant_drugs_count >= 5:
            self.polypharmacy_flag = 1
        return self


def validate_patient_data(
    data: Dict[str, Any],
) -> tuple[bool, Optional[str], Optional[PatientDataSchema]]:
    """Validate raw patient payload dictionary using PatientDataSchema"""
    try:
        schema = PatientDataSchema(**data)
        return True, None, schema
    except Exception as e:
        return False, str(e), None


class ReportGenerationSchema(BaseModel):
    """Schema for report generation request validation"""

    patient_data: Dict[str, Any]
    prediction_result: Dict[str, Any]
    patient_name: str = "Patient"
    patient_id: str = ""
    clinician_name: str = "Clinician"


class DrugInteractionRequestSchema(BaseModel):
    """Schema for drug-drug interaction requests"""

    drugs: List[str] = Field(..., min_length=1, description="List of drug names to check")
    patient_data: Optional[Dict[str, Any]] = Field(default=None)


class ChatRequestSchema(BaseModel):
    """Schema for clinical assistant chat inquiries"""

    message: str = Field(..., min_length=1, max_length=2000)
    context: Optional[Dict[str, Any]] = Field(default=None)


class LabInterpretationRequestSchema(BaseModel):
    """Schema for single lab test interpretation requests"""

    test_name: str = Field(..., min_length=1, max_length=50)
    value: float = Field(...)
    sex: str = Field(default="M")


class ReportRequestSchema(BaseModel):
    """Schema for AI narrative report requests"""

    patient_data: Dict[str, Any] = Field(...)
    prediction_result: Dict[str, Any] = Field(...)
    patient_name: str = Field(default="Patient")
    patient_id: str = Field(default="PT-UNKNOWN")
    clinician_name: str = Field(default="Attending Physician")


class DrugInsightsRequestSchema(BaseModel):
    """Schema for deep AI drug interaction insight requests"""

    medications: List[Any] = Field(..., min_length=2, description="At least two medications")
    patient_age: Any = Field(default="Not specified")
    comorbidities: List[str] = Field(default_factory=list)


class MedicationAnalysisRequestSchema(BaseModel):
    """Schema for medication therapy management analysis requests"""

    patient_data: Dict[str, Any] = Field(...)
    prediction_result: Dict[str, Any] = Field(default_factory=dict)
    patient_name: str = Field(default="Patient")


class AssessmentSaveSchema(BaseModel):
    """Schema for saving complete patient risk assessments"""

    patient_id: str = Field(..., min_length=1)
    patient_data: Dict[str, Any] = Field(...)
    prediction: Dict[str, Any] = Field(...)


class CounsellingSubmitSchema(BaseModel):
    """Schema for patient counselling log submissions"""

    session_type: str = Field(default="Discharge Consultation")
    counsellor_name: str = Field(default="Clinical Pharmacist")
    comprehension_notes: str = Field(default="")
    patient_data: Optional[Dict[str, Any]] = Field(default=None)

File: client/src/lib/schemas.js
import { z } from 'zod'

/**
 * Client-side mirror of PatientDataSchema in schemas.py.
 *
 * Every bound here matches a Pydantic Field(ge=..., le=...) on the server. Keeping
 * them in sync is what turns an opaque 422 into an inline field error. If you change
 * a bound in schemas.py, change it here too.
 *
 * Select option lists mirror the accepted keys in services/pharmacogenomics/* --
 * an unlisted genotype silently scores as "normal activity" on the server, so the
 * UI must not let clinicians invent values.
 */

// Coerces '' -> undefined so an untouched optional number field falls back to its
// default rather than failing as NaN.
const num = (min, max) =>
  z.preprocess(
    (v) => (v === '' || v === null || v === undefined ? undefined : Number(v)),
    z.number({ error: 'Must be a number' }).min(min).max(max),
  )

const int = (min, max) =>
  z.preprocess(
    (v) => (v === '' || v === null || v === undefined ? undefined : Number(v)),
    z.number({ error: 'Must be a number' }).int('Must be a whole number').min(min).max(max),
  )

// Comorbidity / risk flags are 0|1 ints on the server but checkboxes in the UI.
const flag = z.preprocess((v) => (v === true ? 1 : v === false ? 0 : Number(v ?? 0)), z.union([z.literal(0), z.literal(1)]))

export const SEX_OPTIONS = [
  { value: 'M', label: 'Male' },
  { value: 'F', label: 'Female' },
]

export const ETHNICITY_OPTIONS = ['Asian', 'White', 'Black', 'Hispanic', 'Middle Eastern', 'Other']

export const CYP_OPTIONS = {
  cyp2c9: ['Poor', 'Intermediate', 'Wild', 'Rapid'],
  cyp2d6: ['PM', 'IM', 'EM', 'UM'],
  cyp3a4: ['Poor', 'Intermediate', 'Normal', 'Rapid'],
  cyp1a2: ['Slow', 'Intermediate', 'Normal', 'Rapid'],
  cyp2b6: ['Poor', 'Intermediate', 'Normal', 'Rapid'],
  cyp2c19: ['PM', 'IM', 'EM', 'RM', 'UM'],
}

export const CYP_LABELS = {
  cyp2c9: 'CYP2C9',
  cyp2d6: 'CYP2D6',
  cyp3a4: 'CYP3A4',
  cyp1a2: 'CYP1A2',
  cyp2b6: 'CYP2B6',
  cyp2c19: 'CYP2C19',
}

export const TRANSPORTER_OPTIONS = {
  slco1b1_genotype: ['*1/*1', '*1/*5', '*1/*15', '*5/*5', '*15/*15'],
  abcb1_genotype: ['CC', 'CT', 'TT'],
  abcg2_genotype: ['Wild/Wild', 'Wild/Variant', 'Variant/Variant'],
}

export const TRANSPORTER_LABELS = {
  slco1b1_genotype: 'SLCO1B1',
  abcb1_genotype: 'ABCB1',
  abcg2_genotype: 'ABCG2',
}

// Alleles the HLAEngine actually screens for.
export const HLA_OPTIONS = {
  hla_a_typing: ['HLA-A*3101'],
  hla_b_typing: ['HLA-B*5701', 'HLA-B*5801'],
  hla_drb1_typing: ['HLA-DRB1*0701'],
}

export const COMORBIDITY_FIELDS = [
  { name: 'diabetes', label: 'Diabetes mellitus' },
  { name: 'liver_disease', label: 'Liver disease' },
  { name: 'ckd', label: 'Chronic kidney disease' },
  { name: 'cardiac_disease', label: 'Cardiac disease' },
  { name: 'hypertension', label: 'Hypertension' },
  { name: 'respiratory_disease', label: 'Respiratory disease' },
  { name: 'neurological_disease', label: 'Neurological disease' },
  { name: 'autoimmune_disease', label: 'Autoimmune disease' },
]

export const RISK_FLAG_FIELDS = [
  { name: 'cyp_inhibitors_flag', label: 'On CYP inhibitors', hint: 'Concurrent strong CYP450 inhibitor' },
  { name: 'qt_prolonging_flag', label: 'QT-prolonging agent', hint: 'Concurrent QT-prolonging drug' },
  { name: 'hla_risk_allele_flag', label: 'HLA risk allele present', hint: 'Known SCAR-associated allele' },
  { name: 'inpatient_flag', label: 'Inpatient', hint: 'Currently admitted' },
  { name: 'prior_adr_history', label: 'Prior ADR history', hint: 'Documented previous adverse reaction' },
  { name: 'polypharmacy_flag', label: 'Polypharmacy', hint: 'Auto-set when concomitant drugs >= 5' },
  { name: 'high_risk_drug_flag', label: 'High-risk drug', hint: 'Narrow therapeutic index agent' },
]

export const patientSchema = z.object({
  // Demographics
  name: z.string().max(100).default('Patient'),
  patient_id: z.string().max(50).default('PT-UNKNOWN'),
  age: int(0, 125),
  sex: z.enum(['M', 'F']).default('M'),
  ethnicity: z.string().max(50).default('Asian'),

  // Anthropometrics
  height: num(30, 250).default(170),
  weight: num(2, 300).default(70),
  bmi: num(0, 80).default(24.2),

  // Renal
  creatinine: num(0.1, 25).default(1.0),
  egfr: num(1, 200).default(90),

  // Hepatic
  ast_alt: num(1, 5000).default(25),
  bilirubin: num(0.1, 50).default(0.8),
  albumin: num(0.5, 7).default(4.2),

  // Complete blood count
  hemoglobin: num(2, 25).default(14),
  hematocrit: num(5, 75).default(42),
  wbc_count: num(100, 100000).default(7000),
  platelet_count: num(5000, 2000000).default(250000),
  rbc_count: num(0.5, 10).default(4.8),

  // Differential (%)
  neutrophils: num(0, 100).default(60),
  lymphocytes: num(0, 100).default(30),
  monocytes: num(0, 100).default(7),
  eosinophils: num(0, 100).default(2),
  basophils: num(0, 100).default(1),

  // RBC indices
  mcv: num(40, 150).default(90),
  mch: num(10, 50).default(30),
  mchc: num(15, 45).default(34),
  rdw: num(5, 40).default(13),

  // Vitals
  bp_systolic: num(40, 260).default(120),
  bp_diastolic: num(30, 180).default(80),
  heart_rate: num(25, 250).default(75),
  temperature: num(30, 45).default(37),

  // Pharmacogenomics - CYP
  cyp2c9: z.enum(CYP_OPTIONS.cyp2c9).default('Wild'),
  cyp2d6: z.enum(CYP_OPTIONS.cyp2d6).default('EM'),
  cyp3a4: z.enum(CYP_OPTIONS.cyp3a4).default('Normal'),
  cyp1a2: z.enum(CYP_OPTIONS.cyp1a2).default('Normal'),
  cyp2b6: z.enum(CYP_OPTIONS.cyp2b6).default('Normal'),
  cyp2c19: z.enum(CYP_OPTIONS.cyp2c19).default('EM'),

  // Pharmacogenomics - transporters
  slco1b1_genotype: z.enum(TRANSPORTER_OPTIONS.slco1b1_genotype).default('*1/*1'),
  abcb1_genotype: z.enum(TRANSPORTER_OPTIONS.abcb1_genotype).default('CC'),
  abcg2_genotype: z.enum(TRANSPORTER_OPTIONS.abcg2_genotype).default('Wild/Wild'),

  // Pharmacogenomics - HLA
  hla_a_typing: z.array(z.string()).default([]),
  hla_b_typing: z.array(z.string()).default([]),
  hla_drb1_typing: z.array(z.string()).default([]),

  // Drug regimen
  medication_name: z.string().min(1, 'Select a medication').max(100).default('Warfarin'),
  index_drug_dose: num(0.01, 10000).default(5),
  time_since_start_days: int(0, 3650).default(7),
  concomitant_drugs_count: int(0, 50).default(0),

  // Comorbidities
  diabetes: flag.default(0),
  liver_disease: flag.default(0),
  ckd: flag.default(0),
  cardiac_disease: flag.default(0),
  hypertension: flag.default(0),
  respiratory_disease: flag.default(0),
  neurological_disease: flag.default(0),
  autoimmune_disease: flag.default(0),

  // Risk flags
  cyp_inhibitors_flag: flag.default(0),
  qt_prolonging_flag: flag.default(0),
  hla_risk_allele_flag: flag.default(0),
  inpatient_flag: flag.default(0),
  prior_adr_history: flag.default(0),
  polypharmacy_flag: flag.default(0),
  high_risk_drug_flag: flag.default(0),

  // Calculated dosage metrics
  cumulative_dose_mg: num(0, Number.MAX_SAFE_INTEGER).default(0),
  dose_density_mg_day: num(0, Number.MAX_SAFE_INTEGER).default(0),

  // Extra medication context
  external_drugs: z.array(z.record(z.string(), z.any())).default([]),
  external_drugs_list: z.array(z.string()).default([]),
  all_medications: z.array(z.string()).default([]),
})

/**
 * Mirrors PatientDataSchema.compute_derived_metrics (schemas.py:125).
 * Applied before submit so the payload the server receives is the one the UI showed.
 */
export function applyDerivedMetrics(values) {
  const next = { ...values }

  if (next.bmi <= 5.0 && next.height > 0) {
    const hM = next.height / 100
    next.bmi = Math.round((next.weight / (hM * hM)) * 10) / 10
  }

  if (next.concomitant_drugs_count >= 5) {
    next.polypharmacy_flag = 1
  }

  return next
}

/** Field defaults, derived from the schema so there is only one source of truth. */
export const patientDefaults = patientSchema.parse({ age: 45 })

/** Groups drive the sticky section rail and per-section completion on /assessment. */
export const ASSESSMENT_SECTIONS = [
  { id: 'demographics', label: 'Demographics' },
  { id: 'laboratory', label: 'Laboratory Values' },
  { id: 'comorbidities', label: 'Comorbidities' },
  { id: 'medication', label: 'Medication' },
  { id: 'pharmacogenomics', label: 'Pharmacogenomics' },
  { id: 'additional', label: 'Additional Clinical Data' },
]
